torch/_functorch/aot_autograd.py

```diff
@@ -908,34 +930,60 @@ def joint_forward_backward(
-            return torch._to_functional_tensor(t, mirror_autograd_meta=True)
+    def flat_fn_with_synthetic_bases_expanded(
+        primals_before_cloning: List[Any], primals_after_cloning: List[Any], maybe_tangents: Optional[List[Any]]
+    ):
+        # This is where we handle the calling convention around synthetic bases.
+        # We need to make sure that we convert any synthetic base arguments into views
+        # *after* we clone inputs for autograd (see below), to preserve the view relationship.
+        primals = unpack_synthetic_bases(primals_after_cloning, synthetic_base_info)
+        assert len(meta.input_info) == len(primals)
+        outs = forward_or_joint(primals_before_cloning, primals, maybe_tangents)
+        return outs
+
+    def flat_fn_no_input_mutations(primals: List[Any], maybe_tangents: Optional[List[Any]]):
+        # When tracing the joint fwd + bwd, making sure to clone any inputs that are mutated first.
+        # We need to ensure that the inputs we pass to autograd.grad() are the *original*
+        # inputs, and not their mutated values.
+        if maybe_tangents is not None:
+            primals_after_cloning = [
+                maybe_to_fresh_input(i, t) for i, t in enumerate(primals)
+            ]
         else:
-            return t
-
-    def from_fun(t):
-        if not isinstance(t, Tensor) or not torch._is_functional_tensor(t):
-            return t
-        torch._sync(t)
-        return torch._from_functional_tensor(t)
-
-    def functionalized_joint(
-        primals: List[Any], tangents: List[Any]
-    ) -> Tuple[List[Any], List[Any]]:
+            primals_after_cloning = primals
+        outs = flat_fn_with_synthetic_bases_expanded(primals, primals_after_cloning, maybe_tangents)
+        return outs
 
+    def functionalized_f_helper(primals_and_maybe_tangents):
+        if trace_joint:
+            primals, maybe_tangents = primals_and_maybe_tangents
+        else:
+            primals = primals_and_maybe_tangents
+            maybe_tangents = None
         # Wrap inputs into functional wrappers
-        f_primals, f_tangents = pytree.tree_map(to_fun, (primals, tangents))
+        f_primals = pytree.tree_map(to_fun, primals)
+        f_tangents = None if maybe_tangents is None else pytree.tree_map(to_fun, maybe_tangents)
         torch._enable_functionalization(reapply_views=True)
         try:
             # Run the joint
-            f_outs = joint_forward_backward(f_primals, f_tangents)
+            f_outs = flat_fn_no_input_mutations(f_primals, f_tangents)
         finally:
             torch._disable_functionalization()
 
         return pytree.tree_map(from_fun, f_outs)
 
-    return functionalized_joint
+    # the joint needs have args named "primals" and "tangents",
+    # which are hardcoded into the partitioning logic.
+    def traced_joint(primals, tangents):
+        return functionalized_f_helper((primals, tangents))
+
+    def traced_forward(*primals):
```

Comment:
Wouldn't it be more symmetric to just have this be `primals` (no asterisk?)

Reply:
Yep this took me a bit to untangle, I should add a comment.

The main reason it's currently set up this way is because of the calling convention by the partition. The partition takes in a graph that takes in two arguments `(primals, tangents)`, and emits two graphs, that take in the **unpacked** forward args (and backward args, respectively). So in order to keep the `runtime_wrapper` logic consistent across the autograd case and the inference case, I thought it would be simpler if the final compiled function that we call in both cases has the same convention: it takes in the single arg corresponding to "list of input args", and the calls a compiled fn that expects the unpacked args.